src/main/java/org/jenkinsci/remoting/engine/JnlpProtocol4Handler.java

```diff
@@ -0,0 +1,384 @@
+                .named(String.format("%s connection from %s", getName(), socket.getRemoteSocketAddress()))
+                .listener(handler)
+                .build(new ChannelApplicationLayer(threadPool, handler))
+                .get();
+    }
+
+    /**
+     * Handles an outgoing connection to the server on the supplied socket.
+     *
+     * @param socket    the socket.
+     * @param headers   the headers to send to the server.
+     * @param listeners the listeners to process the connection.
+     */
+    @Nonnull
+    @Override
+    public Future<Channel> connect(@Nonnull Socket socket, @Nonnull Map<String, String> headers,
+                                   @Nonnull List<? extends JnlpConnectionStateListener> listeners) throws IOException {
+        NetworkLayer networkLayer = createNetworkLayer(socket);
+        SSLEngine sslEngine = createSSLEngine(socket);
+        sslEngine.setUseClientMode(true);
+        Handler handler = new Handler(createConnectionState(socket, listeners));
+        return ProtocolStack.on(networkLayer)
+                .filter(new AgentProtocolClientFilterLayer(getName()))
+                .filter(new AckFilterLayer())
+                .filter(new SSLEngineFilterLayer(sslEngine, handler))
+                .filter(new ConnectionHeadersFilterLayer(headers, handler))
+                .named(String.format("%s connection to %s", getName(), socket.getRemoteSocketAddress()))
+                .listener(handler)
+                .build(new ChannelApplicationLayer(threadPool, handler))
+                .get();
+    }
+
+    /**
+     * Creates the best network layer for the provided {@link Socket}.
+     *
+     * @param socket the {@link Socket}.
+     * @return the best {@link NetworkLayer} for the provided {@link Socket}.
+     * @throws IOException if the socket is closed already.
+     */
+    private NetworkLayer createNetworkLayer(Socket socket) throws IOException {
+        NetworkLayer networkLayer;
+        SocketChannel socketChannel = isPreferNio() ? socket.getChannel() : null;
+        if (socketChannel == null) {
+            networkLayer = new BIONetworkLayer(ioHub, Channels.newChannel(SocketChannelStream.in(socket)),
+                    Channels.newChannel(SocketChannelStream.out(socket)));
+        } else {
+            networkLayer = new NIONetworkLayer(ioHub, socketChannel, socketChannel);
+        }
+        return networkLayer;
+    }
+
+    /**
+     * Creates an {@link SSLEngine} for the specified {@link Socket}. The {@link SSLContext} may have retained session
+     * state and thus enable a shorter handshake.
+     *
+     * @param socket the socket.
+     * @return the {@link SSLEngine}.                                             s
+     */
+    private SSLEngine createSSLEngine(Socket socket) {
+        SocketAddress remoteSocketAddress = socket.getRemoteSocketAddress();
+        if (remoteSocketAddress instanceof InetSocketAddress) {
```

Comment:
if it is null then we do not have the information to short-cut the handshaking, so the else branch is the correct behaviour